### PR TITLE
[RISCV] Macro-fusion support for veyron-v1 CPU.

### PR DESCRIPTION
**Pull request by mgudim**
Support was added for the following fusions:
  auipc-addi, slli-srli, ld-add
Some parts of the code became repetative, so small refactoring of existing lui-addi fusion was done.

### CODE REVIEW

**Comment by wangpc-pp** [comment]
As far as I know, there may be up to 30+ instruction pairs that can be fused for some macroarchitecture (for example, , [Xiangshan - Decode Unit, Chinese](https://xiangshan-doc.readthedocs.io/zh_CN/latest/frontend/decode/)). If we define subtarget features for all of them, that would be a disaster I think. So, I'm working on a TableGen backend that generates predicators for macro fusion (with some refactor of `ScheduleModel`) and I will raise a RFC and PR later. This is still WIP and  discussions are welcome!

Anyway, this PR looks great to me, and I'm just sharing some information here.

(For current implementation, I don't know if we can refer to PPC's, in which there is a script-generated header.)

**Comment by mgudim** [comment]
> As far as I know, there may be up to 30+ instruction pairs that can be fused for some macroarchitecture (for example, , [Xiangshan - Decode Unit, Chinese](https://xiangshan-doc.readthedocs.io/zh_CN/latest/frontend/decode/)). If we define subtarget features for all of them, that would be a disaster I think. So, I'm working on a TableGen backend that generates predicators for macro fusion (with some refactors of `ScheduleModel`) and I will raise a RFC and PR later. This is still WIP and discussions are welcome!
> 
> Anyway, this PR looks great to me, and I'm just sharing some information here.
> 
> (For current implementation, I don't know if we can refer to PPC's, in which there is a script-generated header.)

Thanks for taking a look. I agree that we shouldn't have too many features defined. I'm happy to rework this in the future if we adopt a new approach.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp [hunk 1]
```diff
@@ -18,6 +18,90 @@
...
+  if (!FirstMI)
+    return true;
+
+  if (FirstMI->getOpcode() != RISCV::ADD)
+    return true;
+
+  return checkRegisters(FirstMI->getOperand(0).getReg(), SecondMI);
+}
+
+// Fuse SLLI by 32 feeding into SRLI by 32 or less or
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:56]: Can we split this into two separate sentences? The `32 or less or` is kind of hard to parse.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp [hunk 2]
```diff
@@ -27,29 +111,18 @@ static bool isLUIADDI(const MachineInstr *FirstMI,
...
     return true;
 
   if (FirstMI->getOpcode() != RISCV::LUI)
     return false;
 
-  Register FirstDest = FirstMI->getOperand(0).getReg();
-
-  // Destination of LUI should be the ADDI(W) source register.
-  if (SecondMI.getOperand(1).getReg() != FirstDest)
+  // The first operand of ADDI might be a frame index.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:121]: This check was removed previously because it wasn't needed. Was this a bad merge or is it needed now?

Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:121]: In this code:

```
$rd0 =  lui $imm0
$rd1 = addi $rs0, $imm1
```

In order for fusion to happen it must be that `$rd0 == rd1` and `$rd0 == rs0`. Same for other fusions. Both checks are done in `checkRegisters` now. I think both checks were needed before two, according to this comment:

```
// Fuse LUI followed by ADDI or ADDIW.
// rd = imm[31:0] which decomposes to
// lui rd, imm[31:12]
// addi(w) rd, rd, imm[11:0]

```

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:121]: I was specifically refering to the FrameIndex comment.

Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:121]: I think we do need the check `SecondMI.getOperand(1).isReg()`  because later we do `SecondMI.getOperand(1).getReg() != FirstDest` - this may crash, right?

I've put that check as a first line of `checkRegisters`.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:121]: As was explained in #68701, if the ADDI operand isn't a register then there can be no data dependency between the LUI and the ADDi so it wouldn't be considered for macrofusion.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp (hunk 1 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp]: I've replaced this with pseudocode.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp (hunk 2 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp]: oh, sorry I didn't realize that by the time we call `shouldScheduleAdjacent` we know that there is a data dependency. Ok, I looked at  `MacroFusion.cpp : 177` - we only consider pairs of vertices connected by an edge of a schedule dag. But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

File: llvm/lib/Target/RISCV/RISCVProcessors.td [hunk 3]
```diff
@@ -242,4 +242,5 @@ def VENTANA_VEYRON_V1 : RISCVProcessorModel<"veyron-v1",
                                              FeatureStdExtZicbom,
                                              FeatureStdExtZicbop,
                                              FeatureStdExtZicboz,
-                                             FeatureVendorXVentanaCondOps]>;
+                                             FeatureVendorXVentanaCondOps,
+                                             TuneVeyronFusions]>;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVProcessors.td:246]: It looks like you're adding this to the feature list, not the tune list.  That's probably not what you meant.

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 4]
```diff
@@ -954,6 +964,14 @@ def TuneSiFive7 : SubtargetFeature<"sifive7", "RISCVProcFamily", "SiFive7",
                                    [TuneNoDefaultUnroll,
                                     TuneShortForwardBranchOpt]>;
 
+def TuneVeyronFusions : SubtargetFeature<"ventana-veyron", "RISCVProcFamily", "VentanaVeyron",
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:967]: Please separate this out in a separate PR which only adds the proc family and the existing LUI/ADDI fusion.  Then this change can extend the list with the new fusions.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp [hunk 5]
```diff
@@ -18,6 +18,97 @@
...
+static bool isSLLISRLI(const MachineInstr *FirstMI,
+                       const MachineInstr &SecondMI) {
+  if (SecondMI.getOpcode() != RISCV::SRLI)
+    return false;
+
+  if (!SecondMI.getOperand(2).isImm())
+    return false;
+
+  unsigned SRLIImm = SecondMI.getOperand(2).getImm();
+  bool IsShiftBy48 = SRLIImm == 48;
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:78]: This seems a very oddly restricted fusion.  Are you sure that e.g. a shift pair by 37 that clears the top bits isn't fused?

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp [hunk 6]
```diff
@@ -18,6 +18,97 @@
...
+
+  // Given SecondMI, when FirstMI is unspecified, we must return
+  // if SecondMI may be part of a fused pair at all.
+  if (!FirstMI)
+    return true;
+
+  if (FirstMI->getOpcode() != RISCV::SLLI)
+    return false;
+
+  unsigned SLLIImm = FirstMI->getOperand(2).getImm();
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:90]: Are intending to allow two different shift amounts less than 32?  That seems unlikely.

**Review by preames — COMMENTED** [review]
Is there an optimization guide or other public reference which describes these fusions?

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp (hunk 5 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:78]: The two shifts by 48 case is the i16 zext pattern for rv64 without Zbb. So it makes some sense.

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 7]
```diff
@@ -937,6 +937,16 @@ def TuneLUIADDIFusion
     : SubtargetFeature<"lui-addi-fusion", "HasLUIADDIFusion",
                        "true", "Enable LUI+ADDI macrofusion">;
 
+def TuneAUIPCADDIFusion
+    : SubtargetFeature<"auipc-addi-fusion", "HasAUIPCADDIFusion",
+                       "true", "Enable AUIPC+ADDI macrofusion">;
+def TuneSLLISRLIFusion
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:943]: This is pretty generic name given the specific cases that are recognized for fusion.

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp (hunk 2 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp]: > But now I am starting to wonder: can we guarantee that an edge always implies data dependency? What if some other mutation creates an artificial edge without data dependency for example?

Yes, I have to say that it's possible to have other order dependencies like `Artificial`. For example, `BaseMemOpClusterMutation::clusterNeighboringMemOps`. Though we may not add artificial edges (not for fusion) between ALU instructions. I haven't run into any problem after remove the `isReg()` yet, but I think we may add it back since there are some potential risks of assertion theoretically.

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 4 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:967]: done: https://github.com/llvm/llvm-project/pull/70414

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp (hunk 6 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:90]: Yes, you're right, thanks. Fixed that.

File: llvm/lib/Target/RISCV/RISCVProcessors.td (hunk 3 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVProcessors.td:246]: done in https://github.com/llvm/llvm-project/pull/70414

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 7 above)
Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:943]: how about `ShiftedZExtFusion`?

**Comment by asb** [comment]
Could we add tests for this please? I know we have macro-fusion-lui-addi.ll already though I'm not sure .ll tests are the best for this. PowerPC's macro-fusion.mir test seemed like a nice way to handle it, to my eye at least.

**Comment by mgudim** [comment]
> Could we add tests for this please? I know we have macro-fusion-lui-addi.ll already though I'm not sure .ll tests are the best for this. PowerPC's macro-fusion.mir test seemed like a nice way to handle it, to my eye at least.

Yes, good idea. Done.

**Comment by mgudim** [comment]
@asb @preames @topperc Ping

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 7 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:966]: ok

File: llvm/test/CodeGen/RISCV/macro-fusions-veyron-v1.mir [hunk 8]
```diff
@@ -0,0 +1,159 @@
+# REQUIRES: asserts
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/macro-fusions-veyron-v1.mir:1]: Do we need end-to-end llc tests like `llvm/test/CodeGen/RISCV/macro-fusion-lui-addi.ll`?

**Review by wangpc-pp — COMMENTED** [review]
I'd like to see this landed so that there are more use cases for TableGen way. :-)

**Comment by mgudim** [comment]
@topperc @asb @preames I think all the comments are addressed, sorry for taking so long. Is it OK to merge?

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/lib/Target/RISCV/RISCVMacroFusion.cpp [hunk 9]
```diff
@@ -18,6 +18,101 @@
...
+  }
+
+  return SecondMI.getOperand(0).getReg() == FirstDest;
+}
+
+// Fuse load with add:
+// add rd, rs1, rs2
+// ld rd, 0(rd)
+static bool isLDADD(const MachineInstr *FirstMI, const MachineInstr &SecondMI) {
+  if (SecondMI.getOpcode() != RISCV::LD)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:41]: @mgudim Is this fusion reall restricted to just 64-bit load?

Comment (mgudim) [reply llvm/lib/Target/RISCV/RISCVMacroFusion.cpp:41]: yes.
